Adds a page for @scroll-timeline rule (#12384)
mdn/content: files/en-us/web/css/@scroll-timeline/index.md

---
title: '@scroll-timeline'
slug: Web/CSS/@scroll-timeline
tags:
  - Animations
  - Scroll
  - At-rule
  - CSS
  - Reference
browser-compat: css.at-rules.scroll-timeline
---
{{CSSRef}}

> The **`@scroll-timeline`** CSS [at-rule](/en-US/docs/Web/CSS/At-rule) defines an [`AnimationTimeline`](/en-US/docs/Web/API/AnimationTimeline) whose time values are determined by progression of scrolling within a scroll container and not by minutes or seconds. Once specified, a scroll timeline is associated with a [CSS Animation](/en-US/docs/Web/CSS/CSS_Animations) by using the `animation-timeline` property.

Reviewer: For symmetry with the release notes edits I suggested

```suggestion
The **`@scroll-timeline`** CSS [at-rule](/en-US/docs/Web/CSS/At-rule) defines an [`AnimationTimeline`](/en-US/docs/Web/API/AnimationTimeline) whose time values are determined by scrolling progress within a scroll container and not by minutes or seconds. Once specified, a scroll timeline is associated with a [CSS Animation](/en-US/docs/Web/CSS/CSS_Animations) by using the `animation-timeline` property.
```

## Syntax

```css
@scroll-timeline moveTimeline {

>   source:  auto;
>   orientation: vertical;
>   scroll-offsets:  0px, 500px;

Reviewer: More consistent spacing

```suggestion
  source: auto;
  orientation: vertical;
  scroll-offsets: 0px, 500px;
```

}
```

### Values

- {{cssxref("custom-ident")}}

>   - : A name identifying the scroll timeline. This must match the identifier production in CSS syntax.

Reviewer: > This must match the identifier production in CSS syntax.

I'm not sure what this means. This sounds like specification speak to me. I'm guessing this is something about what characters names can have. Can you rewrite this more directly?

- `source`
  - : The scrollable element which scrolling position drives the progress of the timeline. Can be:
    - `auto` the `Document` associated with the current global [Window object](/en-US/docs/Web/API/Window).

>     - `selector(“id-selector”)` the scroll container identified by an element's id.

Reviewer: Are the fancy quotes needed or should they be plain `"`?

>     - `selector(“id-selector”)` the scroll container identified by an element's id.

Reviewer: Can these inner lists also be definition lists? I'm imagining something more like the list https://developer.mozilla.org/en-US/docs/Web/CSS/border-style#values where it introduces a list of values and they each get their own definition

Author: This is the only one that's a pain in the butt - if you can spot why the `orientation` list still wants to display as a list, you get a gold star ⭐

Reviewer: OK, that was harder to figure out than I thought. Apparently, you _must_ use spaces, not tabs for this to work (no, I don't know why either). I'd suggest running prettier against your source file: `npx prettier --write files/en-us/web/css/@scroll-timeline/index.md`. That should get rid of some trailing white space and other oddities too.

Author: Sorted - thank you!

    - `none` no scroll container specified.
- `orientation`
  - : The scroll timeline’s orientation:
	  - `auto` defaults to `vertical`
	  - `block` uses the scroll position along the block axis, conforming to writing mode and directionality.
	  - `inline` uses the scroll position along the inline axis, conforming to writing mode and directionality.
	  - `horizontal` uses the horizontal scroll position, regardless of writing mode or directionality.
	  - `vertical` uses the vertical scroll position, regardless of writing mode or directionality.
- `scroll-offsets` 
  - : Determines the scroll timeline's scroll offsets
    - `none` no scroll offsets specified.
    - `<length-pecentage>` A comma separated list of [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage) values.
    - `<element-offset>` An elements position determines the scroll offset.

## Description

> To use the scroll-timeline, create a `@scroll-timeline` rule which is used by the {{cssxref("animation-timeline")}} property to match an animations timeline to it's timeline declaration.

Reviewer: ```suggestion
To use the scroll-timeline, create a `@scroll-timeline` rule, which is used by the {{cssxref("animation-timeline")}} property to match an animations timeline to it's timeline declaration.
```

Each `@scroll-timeline` rule includes properties to determine the source, orientation and scroll-offsets of the scroll timeline. 

### Scroll offsets

> The `scroll-offset` property determines ... This can be set to `none` which means no `scroll-offset` is specified. Or it can be determined by a comma separated list of `<length-percentage>` values. Each value is mapped against the {{cssxref('animation-duration')}}. For instance, if an `animation-duration` is set to `2s` and the scroll offset to `0px, 30px, 100px`, then at 1s the scroll offset would be 30px. Typically you'd use two values here for a smooth scroll animation, such as `0px, 100px`.

Reviewer: > …

Looks like something went unfinished here?

Reviewer: Also this paragraph and the paragraph after suggests that maybe this would be more neatly presented as a list. Something like…

> … can be set one of three ways:
>
> 1. `none`, which means…
> 2. A comma-separate list…
> 3. [Something about element offsets]

The third way of determining a scroll offset is to use an element offset. This means you specify elements within the page, the locations of which determine the scroll timeline and which edge of these elements to use. Specifying elements is done using the `selector()` function, which receives an element's id. Edge is determined by keywords `start` or `end`. An optional threshold value between `0-1` can be used to represent the percentage of the element expected to be visible in the `source`.

```css
@scroll-timeline element-move {
  source: auto;
  orientation: vertical;
  scroll-offsets:
    selector(#myElement) start 0,
    selector(#myElement) end 0
  ;
}
```

## Formal syntax

{{csssyntax}}

## Examples

### Simple example

> 

Reviewer: It'd be nice to introduce this with a description of what's supposed to happen, before narrating its construction. Something like, "In this example, a square rotates as its container scrolls" (or whatever is supposed to happen—it didn't actually work for me)

Author: Yeh unless you're in chrome or FF beta it won't yet

---
title: "@scroll-timeline"
slug: Web/CSS/@scroll-timeline
tags:
  - Animations
  - Scroll
  - At-rule
  - CSS
  - Reference
browser-compat: css.at-rules.scroll-timeline
---

{{CSSRef}}

The **`@scroll-timeline`** CSS [at-rule](/en-US/docs/Web/CSS/At-rule) defines an [`AnimationTimeline`](/en-US/docs/Web/API/AnimationTimeline) whose time values are determined by scrolling progress within a scroll container and not by minutes or seconds. Once specified, a scroll timeline is associated with a [CSS Animation](/en-US/docs/Web/CSS/CSS_Animations) by using the `animation-timeline` property.

## Syntax

```css
@scroll-timeline moveTimeline {
  source: auto;
  orientation: vertical;
  scroll-offsets: 0px, 500px;
}
```

### Values

- {{cssxref("custom-ident")}}

  - : A name identifying the scroll timeline. This name is used when specifying the scroll timeline with the [`animation-timeline`](/en-US/docs/Web/CSS/animation-timeline) property.

- `source`

>   - : The scrollable element which scrolling position drives the progress of the timeline. Can be:

Reviewer 2: ```suggestion
  - : The scrollable element whose scrolling position drives the progress of the timeline. Can be:
```

    - `auto`

      - : The `Document` associated with the current global [Window object](/en-US/docs/Web/API/Window).

>     - `selector("id-selector")`
> 
>       - : the scroll container identified by an element's id.
> 
>     - `none`
>       - : no scroll container specified.

Reviewer 2: ```suggestion
    - `selector("id-selector")`

      - : The scroll container identified by an element's id.

    - `none`
      - : No scroll container specified.
```

- `orientation`

  - : The scroll timeline’s orientation:

    - `auto`

      - : Defaults to `vertical`

    - `block`

      - : Uses the scroll position along the block axis, conforming to writing mode and directionality.

    - `inline`

      - : Uses the scroll position along the inline axis, conforming to writing mode and directionality.

    - `horizontal`

      - : Uses the horizontal scroll position, regardless of writing mode or directionality.

    - `vertical`
      - : Uses the vertical scroll position, regardless of writing mode or directionality.

- `scroll-offsets`

>   - : Determines the scroll timeline's scroll offsets
> 
>     - `none`
>       - : No scroll offsets specified.
>     - `<length-pecentage>`
>       - : A comma separated list of [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage) values.
>     - `<element-offset>`
>       - : An elements position determines the scroll offset.
> 

Reviewer 2: ```suggestion
  - : Determines the scroll timeline's scroll offsets:

    - `none`
      - : No scroll offsets specified.
    - `<length-pecentage>`
      - : A comma separated list of [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage) values.
    - `<element-offset>`
      - : An element’s position determines the scroll offset.

```

## Description

> To use the scroll-timeline, create a `@scroll-timeline` rule, which is used by the {{cssxref("animation-timeline")}} property to match an animations timeline to it's timeline declaration.

Reviewer 2: ```suggestion
To use the scroll timeline, create a `@scroll-timeline` rule, which is used by the {{cssxref("animation-timeline")}} property to match an animation’s timeline to its timeline declaration.
```

Each `@scroll-timeline` rule includes properties to determine the source, orientation and scroll-offsets of the scroll timeline.

### Scroll offsets

The `scroll-offset` property determines where, within the scrolling, the animation should occur. This can be set in one of three ways:

1. By using the CSS keyword `none`, which means no `scroll-offset` is specified.

> 2. It can be determined by a comma separated list of [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage) values. Each value is mapped against the {{cssxref('animation-duration')}}. For instance, if an `animation-duration` is set to `2s` and the scroll offset to `0px, 30px, 100px`, then at 1s the scroll offset would be 30px. Typically you'd use two values here for a smooth scroll animation, such as `0px, 100px`.
> 
> 3. The third way of determining a scroll offset is to use an element offset. This means you specify elements within the page, the locations of which determine the scroll timeline and which edge of these elements to use. Specifying elements is done using the `selector()` function, which receives an element's id. Edge is determined by keywords `start` or `end`. An optional threshold value between `0-1` can be used to represent the percentage of the element expected to be visible in the `source`.

Reviewer 2: ```suggestion
2. It can be determined by a comma-separated list of [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage) values. Each value is mapped against the {{cssxref('animation-duration')}}. For instance, if an `animation-duration` is set to `2s` and the scroll offset to `0px, 30px, 100px`, then at 1s the scroll offset would be 30px. Typically, for a smooth scroll animation, you'd use two values here, such as `0px, 100px`.

3. The third way of determining a scroll offset is to use an element offset. This means you specify elements within the page, the locations of which determine the scroll timeline and which edge of these elements to use. Specifying elements is done using the `selector()` function, which receives an element's id. Edge is determined by keywords `start` or `end`. An optional threshold value between `0–1` can be used to represent the percentage of the element expected to be visible in the `source`.
```

```css
@scroll-timeline element-move {
  source: auto;
  orientation: vertical;
  scroll-offsets: selector(#myElement) start 0, selector(#myElement) end 0;
}
```

## Formal syntax

{{csssyntax}}

## Examples

### Simple example

This example shows a square, which rotates as its container is scrolled vertically. We create an element (`#container`) with a fixed height to allow scroll. This is our `source` element.

> Inside this container we create another element (`#square`), which is styled appropriately to look like a square. This element has a rotation animation applied, using the {{cssxref('@keyframes')}} rule and `animation-name` property.

Reviewer 2: ```suggestion
Inside this container, we create another element (`#square`), which is styled appropriately to look like a square. This element has a rotation animation applied, using the {{cssxref('@keyframes')}} rule and `animation-name` property.
```

> We create a `@scroll-timeline` called `squareTimeline`, setting the `source` as the container, the `orientation` as `vertical` and the `scroll-offset` to start at `0px` and end at `300px` (the height of our container). Then we apply this to the square using the `scroll-timeline` property.

Reviewer 2: ```suggestion
We create an `@scroll-timeline` called `squareTimeline`, setting the `source` as the container, the `orientation` as `vertical` and the `scroll-offset` to start at `0px` and end at `300px` (the height of our container). Then we apply this to the square using the `scroll-timeline` property.
```

#### HTML

```html
<div id="container">
  <div id="square"></div>
</div>
```

#### CSS

```css
#container {
  height: 300px;
}

#square {
  background-color: deeppink;
  width: 100px;
  height: 100px;
  margin-top: 100px;
  animation-name: rotateAnimation;
  animation-duration: 3s;
  animation-direction: alternate;
  animation-timeline: squareTimeline;
}

@scroll-timeline squareTimeline {
  source: selector("#container");
  orientation: "vertical";
  scroll-offsets: 0px, 300px;
}

@keyframes rotateAnimation {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
```

#### Result

{{EmbedLiveSample("Simple_example")}}

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

- [CSS animations](/en-US/docs/Web/CSS/CSS_Animations)

> - [Practical use cases for scroll linked animations in CSS with scroll timelines](https://css-tricks.com/practical-use-cases-for-scroll-linked-animations-in-css-with-scroll-timelines/)

Reviewer 2: ```suggestion
- [Practical use cases for scroll-linked animations in CSS with scroll timelines](https://css-tricks.com/practical-use-cases-for-scroll-linked-animations-in-css-with-scroll-timelines/)
```
